File: android/ssabree/app/src/main/java/com/ssafy/ssabree/features/group/screen/GroupViewModel.kt
package com.ssafy.ssabree.features.group.screen

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.ssafy.ssabree.core.repository.CampusRepository
import com.ssafy.ssabree.core.repository.GroupRepository
import com.ssafy.ssabree.core.repository.KeywordRepository
import com.ssafy.ssabree.core.repository.MemberRepository
import com.ssafy.ssabree.features.group.model.GroupKind
import com.ssafy.ssabree.features.group.model.GroupListItemUiModel
import com.ssafy.ssabree.features.group.model.toUiModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.launch
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Locale

data class GroupUiState(
    val groupKind: GroupKind,
    val selectedFilter: String = "전체",
    val groups: List<GroupListItemUiModel> = emptyList(),
    val recentKeywords: List<String> = emptyList(),
    val campusId: Long? = null,
    val showMinLengthError: Boolean = false,
    val isLoading: Boolean = false,
    val errorMessage: String? = null
) {
    val filteredGroups: List<GroupListItemUiModel> =
        if (selectedFilter == "전체") {
            groups
        } else {
            groups.filter { it.categoryLabel == selectedFilter }
        }
}

class GroupViewModel(
    private val groupRepository: GroupRepository,
    private val keywordRepository: KeywordRepository,
    private val memberRepository: MemberRepository,
    private val campusRepository: CampusRepository,
    private val groupKind: GroupKind
) : ViewModel() {

    private val _uiState = MutableStateFlow(GroupUiState(groupKind = groupKind))
    val uiState: StateFlow<GroupUiState> = _uiState.asStateFlow()
    init {
        load()
        loadRecentKeywords()
    }

    fun load() = viewModelScope.launch {
        _uiState.update { it.copy(isLoading = true, errorMessage = null) }
        val campusId = _uiState.value.campusId ?: resolveCampusId().also { resolved ->
            _uiState.update { it.copy(campusId = resolved) }
        }
        val result = if (groupKind == GroupKind.STUDY) {
            groupRepository.getStudies(campusId = campusId)
        } else {
            groupRepository.getTeams(campusId = campusId)
        }
        result.onSuccess { items ->
            val uiItems = coroutineScope {
                items.map { model ->
                    async {
                        val membersResult = if (groupKind == GroupKind.STUDY) {
                            groupRepository.getStudyMembers(model.id)
                        } else {
                            groupRepository.getTeamMembers(model.id)
                        }
                        val currentMembers = membersResult.getOrNull()?.size
                            ?: (model.currentMembers ?: 0)
                        model.toUiModel(groupKind, currentMembers)
                    }
                }.awaitAll()
            }
            _uiState.update {
                it.copy(
                    groups = uiItems,
                    isLoading = false
                )
            }
        }.onFailure { e ->
            _uiState.update { it.copy(isLoading = false, errorMessage = e.message) }
        }
    }

    fun onFilterSelected(filter: String) {
        _uiState.update { it.copy(selectedFilter = filter) }
    }

    private fun loadRecentKeywords() {
        val keywords = keywordRepository.getRecentKeywords()
        _uiState.update { it.copy(recentKeywords = keywords) }
    }

    fun onSearchSubmit(query: String) {
        val keyword = query.trim()
        if (keyword.length < 2) {
            _uiState.update { it.copy(showMinLengthError = true) }
            return
        }
        keywordRepository.saveKeyword(keyword)
        loadRecentKeywords()
        _uiState.update { it.copy(showMinLengthError = false) }
    }

    fun deleteRecentKeyword(keyword: String) {
        keywordRepository.deleteKeyword(keyword)
        loadRecentKeywords()
    }

    fun clearMinLengthError() {
        _uiState.update { it.copy(showMinLengthError = false) }
    }

    private suspend fun resolveCampusId(): Long? {
        val campusName = memberRepository.getMyPage().getOrNull()?.user?.campus?.trim()
        if (campusName.isNullOrBlank()) return null
        val campuses = campusRepository.getCampuses().getOrNull().orEmpty()
        val normalized = normalizeCampusName(campusName)
        return campuses.firstOrNull { campus ->
            val campusNormalized = normalizeCampusName(campus.name)
            campusNormalized == normalized ||
                campusNormalized.contains(normalized) ||
                normalized.contains(campusNormalized)
        }?.id?.toLong()
    }

    private fun normalizeCampusName(name: String): String {
        return name.replace("캠퍼스", "").replace(" ", "").lowercase()
    }

    private fun isExpired(endDate: String): Boolean {
        val input = SimpleDateFormat("yyyy-MM-dd", Locale.KOREA)
        return runCatching {
            val end = input.parse(endDate) ?: return false
            val calendar = Calendar.getInstance().apply {
                set(Calendar.HOUR_OF_DAY, 0)
                set(Calendar.MINUTE, 0)
                set(Calendar.SECOND, 0)
                set(Calendar.MILLISECOND, 0)
            }
            end.before(calendar.time)
        }.getOrDefault(false)
    }
}

File: iOS/ssabree/ssabree/Features/Group/GroupViewModel.swift
import Foundation
import Observation

struct GroupUiState {
    let groupKind: GroupKind
    var selectedFilter: String = "전체"
    var groups: [GroupListItemUiModel] = []
    var recentKeywords: [String] = []
    var campusId: Int? = nil
    var showMinLengthError: Bool = false
    var isLoading: Bool = false
    var errorMessage: String?

    var filteredGroups: [GroupListItemUiModel] {
        if selectedFilter == "전체" {
            return groups
        } else {
            return groups.filter { $0.categoryLabel == selectedFilter }
        }
    }
}

struct GroupListItemUiModel: Identifiable {
    let id: Int
    let title: String
    let categoryLabel: String
    let dDay: String
    let currentMembers: Int
    let maxMembers: Int
    let status: String
    let isExpired: Bool
    let isFull: Bool
    let isClosed: Bool
    let closedMessage: String?
}

@Observable
final class GroupViewModel {
    private let groupRepository: GroupRepository
    private let keywordRepository: KeywordRepository
    private let myPageRepository: MyPageRepository
    private let campusRepository: CampusRepository
    private let groupKind: GroupKind

    var uiState: GroupUiState

    init(
        groupRepository: GroupRepository,
        keywordRepository: KeywordRepository? = nil,
        myPageRepository: MyPageRepository? = nil,
        campusRepository: CampusRepository? = nil,
        initialKind: GroupKind = .study
    ) {
        self.groupRepository = groupRepository
        self.keywordRepository = keywordRepository ?? KeywordRepositoryImpl()
        self.myPageRepository = myPageRepository ?? FakeMyPageRepository()
        self.campusRepository = campusRepository ?? FakeCampusRepository()
        self.groupKind = initialKind
        self.uiState = GroupUiState(groupKind: initialKind)
    }

    @MainActor
    func load() async {
        uiState.isLoading = true
        uiState.errorMessage = nil

        // Resolve campus ID if not already set
        if uiState.campusId == nil {
            uiState.campusId = await resolveCampusId()
        }

        let result: Result<[GroupSummaryModel], Error>
        if groupKind == .study {
            result = await groupRepository.getStudies(campusId: uiState.campusId, type: nil)
        } else {
            result = await groupRepository.getTeams(campusId: uiState.campusId, type: nil)
        }

        switch result {
        case .success(let items):
            // Android와 동일하게 각 그룹의 실제 멤버 수를 별도 API로 조회 (병렬)
            let kind = groupKind
            let repo = groupRepository
            let uiItems = await withTaskGroup(of: (Int, Int).self) { group in
                for model in items {
                    group.addTask {
                        let membersResult: Result<[GroupMemberModel], Error>
                        if kind == .study {
                            membersResult = await repo.getStudyMembers(studyId: model.id)
                        } else {
                            membersResult = await repo.getTeamMembers(teamId: model.id)
                        }
                        let count = (try? membersResult.get())?.count ?? (model.currentMembers ?? 0)
                        return (model.id, count)
                    }
                }

                var memberCounts: [Int: Int] = [:]
                for await (id, count) in group {
                    memberCounts[id] = count
                }
                return memberCounts
            }

            // 활성 그룹 먼저, 비활성(마감/만석) 그룹 나중에 정렬
            let allGroups = items.map { model in
                model.toListItemUiModel(kind: groupKind, currentMembersOverride: uiItems[model.id])
            }
            uiState.groups = allGroups.sorted { a, b in
                if a.isClosed != b.isClosed { return !a.isClosed }
                return false
            }
            uiState.isLoading = false
        case .failure(let error):
            uiState.errorMessage = error.localizedDescription
            uiState.isLoading = false
        }
    }

    func loadRecentKeywords() {
        uiState.recentKeywords = keywordRepository.getRecentKeywords()
    }

    func onFilterSelected(_ filter: String) {
        uiState.selectedFilter = filter
    }

    func onSearchSubmit(_ query: String) {
        let keyword = query.trimmingCharacters(in: .whitespaces)
        if keyword.count < 2 {
            uiState.showMinLengthError = true
            return
        }
        keywordRepository.addRecentKeyword(keyword: keyword)
        loadRecentKeywords()
        uiState.showMinLengthError = false
    }

    func deleteRecentKeyword(_ keyword: String) {
        keywordRepository.deleteRecentKeyword(keyword: keyword)
        loadRecentKeywords()
    }

    func clearMinLengthError() {
        uiState.showMinLengthError = false
    }

    private func resolveCampusId() async -> Int? {
        let myPageResult = await myPageRepository.getMyPage()
        guard case .success(let myPage) = myPageResult,
              let campusName = myPage.user?.campus?.trimmingCharacters(in: .whitespaces),
              !campusName.isEmpty else {
            return nil
        }

        let campusesResult = await campusRepository.getCampuses()
        guard case .success(let campuses) = campusesResult else {
            return nil
        }

        let normalized = normalizeCampusName(campusName)
        return campuses.first { campus in
            let campusNormalized = normalizeCampusName(campus.name)
            return campusNormalized == normalized ||
                   campusNormalized.contains(normalized) ||
                   normalized.contains(campusNormalized)
        }?.id
    }

    private func normalizeCampusName(_ name: String) -> String {
        return name.replacingOccurrences(of: "캠퍼스", with: "")
            .replacingOccurrences(of: " ", with: "")
            .lowercased()
    }
}

// MARK: - GroupSummaryModel Extension

extension GroupSummaryModel {
    func toListItemUiModel(kind: GroupKind, currentMembersOverride: Int? = nil) -> GroupListItemUiModel {
        let categoryLabel = kind == .study
            ? GroupTypeMapper.studyTypeToLabel(type)
            : GroupTypeMapper.teamTypeToLabel(type)

        let dDayText = calculateDDay(endDate: endDate)

        let current = currentMembersOverride ?? currentMembers ?? 0
        let expired = isExpired
        let full = current >= capacity
        let closed = expired || full
        let closedMsg: String? = if expired {
            "모집 기간이 지났습니다."
        } else if full {
            "그룹 인원이 꽉 찼습니다."
        } else {
            nil
        }

        let displayStatus: String
        if let apiStatus = status?.uppercased() {
            switch apiStatus {
            case "OPEN": displayStatus = "모집중"
            case "ONGOING": displayStatus = "진행중"
            case "CLOSED": displayStatus = "완료"
            default: displayStatus = "모집중"
            }
        } else {
            displayStatus = "모집중"
        }

        return GroupListItemUiModel(
            id: id,
            title: title,
            categoryLabel: categoryLabel,
            dDay: dDayText,
            currentMembers: current,
            maxMembers: capacity,
            status: displayStatus,
            isExpired: expired,
            isFull: full,
            isClosed: closed,
            closedMessage: closedMsg
        )
    }

    private func calculateDDay(endDate: String?) -> String {
        guard let endStr = endDate else { return "" }

        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd"
        formatter.locale = Locale(identifier: "ko_KR")

        guard let end = formatter.date(from: endStr) else { return "" }

        let calendar = Calendar.current
        let now = Date()
        let components = calendar.dateComponents([.day], from: now, to: end)

        guard let days = components.day else { return "" }

        if days < 0 {
            return "마감"
        } else if days == 0 {
            return "D-Day"
        } else {
            return "D-\(days)"
        }
    }
}
